MaterialDesignThemes.nuspec

```
@@ -17,12 +17,15 @@
     <dependencies>
       <group targetFramework="net462">
         <dependency id="MaterialDesignColors" version="[1.2.1, 2.0)" />
+        <dependency id="Microsoft.Xaml.Behaviors.Wpf" version="1.1.31" />
```

Sorry I missed your earlier comment about wanting help with the syntax here. I would set it up to allow for a range of value. Assuming the library is following semantic versioning (which it appears they are) we should be good to tell consumers anything from this version up to (but excluding) a 2.0 release. Like this:
```suggestion
        <dependency id="Microsoft.Xaml.Behaviors.Wpf" version="[1.1.31, 2.0)" />
```

---

I was not really clear about what I meant here...

I think your suggestion to limit between a minimum version and a max version is reasonable. I think my original value meant that it should be "equal to or greater than" (ie. no upper limit). If you want a specific version, I think you put a single version in the hard brackets if I remember correctly.

Anyways, my concern was/is this:

We probably don't have a hard requirement on the latest version, we could compile against an older version in order to reach a broader audience (if they are locked into an older version of the same dependency for some reason). As such, I thought we could just put the smallest version in there which we know works.
**However**, my concern is regarding what the nuspec file says vs. what we actually compile against. If those two don't match (which dependabot would make sure was the case if we opted for an early version in the nuspec), I am under the impression that consumers of the library potentially would need a "binding redirect" even if they only have MDIX as a nuget package. I don't know this for sure, but I base the assumption on the fact that installing the nuget package, I would assume the nuget package manager simply pulls in the dependency in the (minimum) version that is spec'ed in the nuspec. This means the user ends up with an older version on disk, but MDIX compiled against (and thus expecting) a newer version; I believe this needs to be mitigated with a binding redirect.

You probably know more than I do about this, so please enlighten me :-)

---

This is a great question. After pulling the docs for it I realized I was wrong to request an upper bound on the version as the [recommendation from MS says to avoid that](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/dependencies#nuget-dependency-version-ranges). I think we can also likely set our NuGet dependency to use a lower bound of `[1.1.3](https://www.nuget.org/packages/Microsoft.Xaml.Behaviors.Wpf/1.1.3)` and no upper bound (this was a mistake on my part with my earlier suggestion).

However on to your other question about how referenced assemblies work. First it is worth mentioning that the way [.NET Framework runtime](https://learn.microsoft.com/dotnet/framework/deployment/how-the-runtime-locates-assemblies) and the [.NET Core Runtime](https://learn.microsoft.com/dotnet/standard/assembly/resolve-loads) resolve dependencies is a little different (largely around the use of the GAC; Global Assembly Cache). It is a common misconception that the version of the assembly that you compile against will be the same version of the assembly that your run against. When loading NuGet dependencies it uses the lower common dependency version. For projects that wish to have a newer version of the library that can explicitly reference it with the version that they would like, as long as this version doesn't cause an error with other packages (or the error is suppressed) the project will then load that NuGet dependency. This then means that NuGet's assemblies (anything under the `lib` directory in the NuGet that matches the target framework) will be copies to the projects output directory (ie. `bin\Debug` or similar). When the project is launched, the runtime then takes over and attempts to resolve the assembly (details in those two links above). However, for most NuGet packages, this ends up simply resolving to the assembly that is alongside the running application (the same one that was copied from the NuGet package). As long as there are no errors resolving the public API members being accessed, it is able to use whatever version of the library it finds.

---

@Keboo Thanks for the explanation. I am seemingly not very good at communicating my concern here :-) Reading the docs, I think they just confirm what I am trying to say (but failing to communicate). I am aware that you don't necessarily run against the same version of the dependent assembly as you compiled against; but it will attempt to bind with the exact version you compiled against unless "told otherwise" (eg. `<bindingRedirect>` in `App.Config` or another configuration file).

Say we have a project, P, which has 2 dependencies: A and B. Both of these assemblies have a dependency on C, but with different minimum versions (A requires min C.v1, and B requires min C.v2). Since A and B are compiled against different versions of C, project P needs to make a decision on which version of C to use. Typically, this ends up being handled by a `<bindingRedirect>`. In this example, the redirect would probably say that "for any requested version of C between v1 and v2, use v2". This would mean that dependency A, when resolving its dependency C will get C.v2 although it was compiled against and therefore is requesting C.v1. This is at least my understanding, and also something I often times need to deal with in my job.

Back to my concern; which is a slight variation of the above.
Say we have project P, which references MDIX. The MDIX nuget package indicates that the min required version of `Microsoft.Xaml.Behaviors.Wpf` is v1.1.3, however MDIX is compiled against v1.1.37 (latest). My concern is here that the nuget package manager will put v1.1.3 in the lib folder and copy that to the bin folder when compiling. To my understanding, the application P then needs a `<bindingRedirect>` in a configuration file in order to resolve `Microsoft.Xaml.Behaviors.Wpf` to v.1.1.3 when requesting v1.1.37 (which it was compiled against); this seems odd when simply pulling in a single nuget package.
This is why I believe we should also compile against the min version we state in the nuspec, and therefore avoid dependabot updates. Aagain, I am not 100% sure which version the nuget package manager pulls down and puts in the lib/bin folders, my assumption is that it is the one stated in the nuspec, but I could be wrong.